### PR TITLE
[SPARK-45747][SS] Use prefix key information in state metadata to handle reading state for session window aggregation

### PR DESCRIPTION
**Pull request by chaoqin-li1123**
### What changes were proposed in this pull request?
Currently reading state for session window aggregation operator is not supported because the numColPrefixKey is unknown. We can read the operator state metadata introduced in SPARK-45558 to determine the number of prefix columns and load the state of session window correctly.

### Why are the changes needed?
To support reading state for session window aggregation.

### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?
Add integration test.

### Was this patch authored or co-authored using generative AI tooling?
No.

### CODE REVIEW

**Comment by chaoqin-li1123** [comment]
@HeartSaVioR

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/StatePartitionReader.scala [hunk 1]
```diff
@@ -57,12 +58,31 @@ class StatePartitionReader(
       partition.sourceOptions.operatorId, partition.partition, partition.sourceOptions.storeName)
     val stateStoreProviderId = StateStoreProviderId(stateStoreId, partition.queryId)
 
+    val allStateStoreMetadata = new StateMetadataPartitionReader(
+      partition.sourceOptions.stateCheckpointLocation.getParent.toString, hadoopConf)
+      .stateMetadata.toArray
+
+    val stateStoreMetadata = allStateStoreMetadata.filter(entry =>
```
Comment (HeartSaVioR) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/StatePartitionReader.scala:65]: nit: `(entry =>` to ` { entry =>` (attention on the space around {)

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/StatePartitionReader.scala [hunk 2]
```diff
@@ -57,12 +58,31 @@ class StatePartitionReader(
...
+        "is created by older version of spark. The state of session window aggregation can't be " +
+        "read correctly without state metadata and runtime exception will be thrown. " +
+        "Run the streaming query in newer spark version to generate state metadata.")
+      0
+    } else {
+      require(stateStoreMetadata.length == 1)
+      stateStoreMetadata.head.numColsPrefixKey
+    }
+
     // TODO: This does not handle the case of session window aggregation; we don't have an
```
Comment (HeartSaVioR) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/StatePartitionReader.scala:80]: Let's remove this TODO comment as you're addressing this.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/metadata/StateMetadataSource.scala [hunk 3]
```diff
@@ -195,7 +195,7 @@ class StateMetadataPartitionReader(
     }
   }
 
-  private lazy val stateMetadata: Iterator[StateMetadataTableEntry] = {
+  private[state] lazy val stateMetadata: Iterator[StateMetadataTableEntry] = {
```
Comment (HeartSaVioR) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/metadata/StateMetadataSource.scala:198]: Is this change necessary?

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/v2/state/StateDataSourceReadSuite.scala [hunk 4]
```diff
@@ -494,6 +497,48 @@ abstract class StateDataSourceReadSuite extends StateDataSourceTestBase with Ass
     }
   }
 
+  test("Session window aggregation") {
+    withTempDir { checkpointDir =>
+      val input = MemoryStream[(String, Long)]
```
Comment (HeartSaVioR) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/v2/state/StateDataSourceReadSuite.scala:502]: Shall we follow the pattern we do for StateDataSourceReadSuite vs StateDataSourceTestBase? The main reason I put the part of query execution to StateDataSourceTestBase is that we'd probably be likely to reuse the query between test for read vs test for write.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/v2/state/StateDataSourceReadSuite.scala [hunk 5]
```diff
@@ -494,6 +497,48 @@ abstract class StateDataSourceReadSuite extends StateDataSourceTestBase with Ass
...
+        CheckNewAnswer(
+          ("hello", 40, 51, 11, 2),
+          ("world", 40, 51, 11, 2),
+          ("streaming", 40, 51, 11, 2),
+          ("spark", 40, 50, 10, 1),
+          ("structured", 41, 51, 10, 1)
+        ),
+        StopStream
+      )
+
```
Comment (HeartSaVioR) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/v2/state/StateDataSourceReadSuite.scala:533]: nit: single empty line

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/v2/state/StateDataSourceReadSuite.scala [hunk 6]
```diff
@@ -494,6 +497,48 @@ abstract class StateDataSourceReadSuite extends StateDataSourceTestBase with Ass
...
+
+      val df = spark.read.format("statestore").load(checkpointDir.toString)
+        .selectExpr("key.sessionId as sessionId", "value.count as count")
```
Comment (HeartSaVioR) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/v2/state/StateDataSourceReadSuite.scala:536]: Shall we validate the full schema of the state, including the part of session window?

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/v2/state/StateDataSourceReadSuite.scala (hunk 4 above)
Comment (chaoqin-li1123) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/v2/state/StateDataSourceReadSuite.scala]: Moved the query running code to StateDataSourceTestBase.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/metadata/StateMetadataSource.scala (hunk 3 above)
Comment (chaoqin-li1123) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/metadata/StateMetadataSource.scala:198]: Yes, I use this class to get prefix key number but don't want to access the internal row because it is untyped. So I need to exposed the api to access the metadata as case class

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/StatePartitionReader.scala (hunk 2 above)
Comment (chaoqin-li1123) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/StatePartitionReader.scala]: Removed.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/StatePartitionReader.scala (hunk 1 above)
Comment (chaoqin-li1123) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/StatePartitionReader.scala]: fixed.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/v2/state/StateDataSourceReadSuite.scala (hunk 6 above)
Comment (chaoqin-li1123) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/v2/state/StateDataSourceReadSuite.scala]: Changed the query to validate all the field.

**Review by HeartSaVioR — APPROVED** [review]
+1 pending CI.

**Comment by HeartSaVioR** [comment]
Thanks! Merging to master.

**Comment by HeartSaVioR** [comment]
@chaoqin-li1123 Could you please rebase your change with latest master branch? merge script is confusing that I'm the main author due to my commits listed here.
